<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Quang Nguyen" />

<meta name="date" content="2022-03-12" />

<title>Preparing trait databases</title>

<script src="site_libs/header-attrs-2.12/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Evaluating trait databases for taxon-set analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Preparation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="retrieve_agp.html">AGP data manifest</a>
    </li>
    <li>
      <a href="agp_download.html">Downloading AGP</a>
    </li>
    <li>
      <a href="db_prep.html">Updating database</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Main Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Coverage analyses</li>
    <li class="dropdown-header">Concordance analyses</li>
    <li class="dropdown-header">Predictive analyses</li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/qpmnguyen/microbe_set_trait">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preparing trait databases</h1>
<h4 class="author">Quang Nguyen</h4>
<h4 class="date">2022-03-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-03-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>microbe_set_trait/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220309code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220309)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220309code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220309)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
unnamed-chunk-4
</li>
<li>
unnamed-chunk-5
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>db_prep_cache</code> and re-run the analysis. To have workflowr
automatically delete the cache directory prior to building the file, set
<code>delete_cache = TRUE</code> when running <code>wflow_build()</code>
or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomqpmnguyenmicrobesettraittree154a960604fe05b408324c5d17e689f6477ff635targetblank154a960a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/qpmnguyen/microbe_set_trait/tree/154a960604fe05b408324c5d17e689f6477ff635" target="_blank">154a960</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomqpmnguyenmicrobesettraittree154a960604fe05b408324c5d17e689f6477ff635targetblank154a960a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/qpmnguyen/microbe_set_trait/tree/154a960604fe05b408324c5d17e689f6477ff635" target="_blank">154a960</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Renviron
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .nextflow.log
    Ignored:    .nextflow.log.1
    Ignored:    .nextflow.log.2
    Ignored:    .nextflow.log.3
    Ignored:    .nextflow.log.4
    Ignored:    .nextflow.log.5
    Ignored:    .nextflow.log.6
    Ignored:    .nextflow/
    Ignored:    R/prototype.R
    Ignored:    _targets/
    Ignored:    analysis/db_prep_cache/
    Ignored:    databases/
    Ignored:    large_files/
    Ignored:    picrust2_tmp/
    Ignored:    python/.ipynb_checkpoints/
    Ignored:    python/data/
    Ignored:    renv/staging/
    Ignored:    store_coverage/
    Ignored:    store_dada2/
    Ignored:    store_db_preprocess/
    Ignored:    store_pred/
    Ignored:    work/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/db_prep.Rmd</code>) and HTML
(<code>docs/db_prep.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/154a960604fe05b408324c5d17e689f6477ff635/analysis/db_prep.Rmd" target="_blank">154a960</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-21
</td>
<td>
workflowr::wflow_publish("analysis/db_prep.Rmd")
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/cf04966d791f29f5ac7d557846db307a7994c5db/analysis/db_prep.Rmd" target="_blank">cf04966</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-20
</td>
<td>
More db prep work
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/8ccd1ff819484e02a61fb51208191b6f51de6673/analysis/db_prep.Rmd" target="_blank">8ccd1ff</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-18
</td>
<td>
Switching dbs
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/624d0bf6b2a27975f814b9c89173a7fb50778167/analysis/db_prep.Rmd" target="_blank">624d0bf</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-17
</td>
<td>
Adding more details to db_prep
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/7ce15a739ff55ff4b9c5bc6ef18a8f2be8fcc09f/analysis/db_prep.Rmd" target="_blank">7ce15a7</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-17
</td>
<td>
new db_prep
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/d42b3f0d3e0de0965406a25a41b972c321cf9b5a/analysis/db_prep.Rmd" target="_blank">d42b3f0</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-16
</td>
<td>
Reorganizng site and adding new files
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/85f49e075cd737c0dc423a64cca3bba12ade531a/analysis/db_prep.Rmd" target="_blank">85f49e0</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-15
</td>
<td>
Still more db prep
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/e3e031d14a1ffc10f92358f5875b62af8b2becfa/analysis/db_prep.Rmd" target="_blank">e3e031d</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-15
</td>
<td>
More dbprep
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/9cc43817da1110e1ae2725df07b9e066f1dcc9f5/analysis/db_prep.Rmd" target="_blank">9cc4381</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-15
</td>
<td>
Starting db_prep
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/qpmnguyen/microbe_set_trait/9cc43817da1110e1ae2725df07b9e066f1dcc9f5/docs/db_prep.html" target="_blank">9cc4381</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-15
</td>
<td>
Starting db_prep
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/7175fe96505114d870bc34e069628e1aa6e259d8/analysis/db_prep.Rmd" target="_blank">7175fe9</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-14
</td>
<td>
Add a skip function
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/qpmnguyen/microbe_set_trait/d785d21db27e033600d31dbc5ea2b4dfac63f60f/docs/db_prep.html" target="_blank">d785d21</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/qpmnguyen/microbe_set_trait/e73f8c0d32289aafbcb30b15c6862f61c042b26a/docs/db_prep.html" target="_blank">e73f8c0</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/909f189e3b47953c57df3cbcb456cec0d0678917/analysis/db_prep.Rmd" target="_blank">909f189</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
workflowr::wflow_publish("analysis/db_prep.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>We’re prepping the latest database for export for evaluation. For
this manuscript, we’re merging a couple of existing databases:</p>
<ol style="list-style-type: decimal">
<li>The comprehensive synthesis of trait-database from <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7275036/">Madin et
al. 2020</a>. This database was last updated in 2020. Most of the
database’s sources are static sources, with the exception of the <a
href="https://gold.jgi.doe.gov/downloads">GOLD database</a>. As such,
we’re merging the existing release of the Madin et al. database with the
most recent GOLD release (2022-03-12).<br />
</li>
<li>Manual curation of bergey’s manual by <a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04216-2">Weissman
et al.</a>. This database contains manual curation of the Bergey’s
manual specific to human-associated microbiomes.</li>
</ol>
<p>The way we’re trying to combine these disparate sources would be to
perform something similar to Madin et al. using the <a
href="https://github.com/bacteria-archaea-traits/bacteria-archaea-traits/blob/master/R/functions.R">R
code</a> on GitHub. We’re going to apply relevant transformations and
mappings where apply.</p>
</div>
<div id="analysis" class="section level1">
<h1>Analysis</h1>
<p>Loading some packages</p>
<pre class="r"><code>library(piggyback)
library(data.table)
library(dtplyr)
library(targets)
library(here)
#&gt; here() starts at /Users/quangnguyen/research/microbe_set_trait
library(stringdist)
library(tidyverse)
#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──
#&gt; ✓ ggplot2 3.3.5     ✓ purrr   0.3.4
#&gt; ✓ tibble  3.1.6     ✓ dplyr   1.0.8
#&gt; ✓ tidyr   1.2.0     ✓ stringr 1.4.0
#&gt; ✓ readr   2.1.2     ✓ forcats 0.5.1
#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
#&gt; x dplyr::between()   masks data.table::between()
#&gt; x tidyr::extract()   masks stringdist::extract()
#&gt; x dplyr::filter()    masks stats::filter()
#&gt; x dplyr::first()     masks data.table::first()
#&gt; x dplyr::lag()       masks stats::lag()
#&gt; x dplyr::last()      masks data.table::last()
#&gt; x purrr::transpose() masks data.table::transpose()
here::i_am(&quot;analysis/db_prep.Rmd&quot;)
#&gt; here() starts at /Users/quangnguyen/research/microbe_set_trait</code></pre>
<div id="downloading-data-files" class="section level2">
<h2>Downloading data files</h2>
<p>Downloading data files from respective sources and version control on
GitHub using <code>piggyback</code></p>
<pre class="r"><code># Not run since file is large and is already uploaded to GitHub. Code here for provenance 
piggyback::pb_upload(file = here(&quot;large_files&quot;, &quot;goldData.xlsx&quot;), tag = &quot;0.1&quot;, overwrite = TRUE)
</code></pre>
</div>
<div id="processing-data-files" class="section level2">
<h2>Processing data files</h2>
<p>First, load the original Madin et al. database</p>
<pre class="r"><code>base &lt;- read_csv(here(&quot;data&quot;, &quot;condensed_species_NCBI.txt&quot;)) %&gt;% 
    select(species_tax_id, superkingdom, phylum, class, order, family, 
           genus, species, metabolism, gram_stain, pathways, 
           carbon_substrates, sporulation, motility, cell_shape) %&gt;% 
    rename(&quot;substrate&quot; = carbon_substrates)
#&gt; Rows: 14893 Columns: 79
#&gt; ── Column specification ────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (19): species, genus, family, order, class, phylum, superkingdom, gram_s...
#&gt; dbl (60): species_tax_id, d1_lo, d1_up, d2_lo, d2_up, doubling_h, genome_siz...
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div id="processing-gold" class="section level3">
<h3>Processing GOLD</h3>
<p>Let’s process GOLD. First, load data from <code>piggyback</code>
(this is cached). Script follows existing processing pipeline <a
href="https://github.com/bacteria-archaea-traits/bacteria-archaea-traits/blob/master/R/preparation/gold.R">here</a>
from Madin et al. </p>
<pre class="r"><code>pth &lt;- here(&quot;large_files&quot;, &quot;goldData.csv&quot;)
if (!file.exists(pth)){
    piggyback::pb_download(file = &quot;goldData.xlsx&quot;, dest = here(&quot;large_files&quot;), tag = &quot;0.1&quot;)
    gold &lt;- readxl::read_xlsx(path = here(&quot;large_files&quot;, &quot;goldData.xlsx&quot;), sheet = &quot;Organism&quot;)
    readr::write_csv(x = gold, file = pth)
}
gold &lt;- read_csv(file = pth)
#&gt; Rows: 428241 Columns: 42
#&gt; ── Column specification ────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (35): ORGANISM GOLD ID, ORGANISM NAME, ORGANISM NCBI SUPERKINGDOM, ORGAN...
#&gt; dbl  (4): ORGANISM NCBI TAX ID, ORGANISM ISOLATION PUBMED ID, ORGANISM ECOSY...
#&gt; lgl  (3): ORGANISM SALINITY CONCENTRATION, ORGANISM PRESSURE, ORGANISM CARBO...
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Let’s perform name conversions</p>
<pre class="r"><code># convert names 
colnames(gold) &lt;- colnames(gold) %&gt;% 
    gsub(x = ., pattern = &quot; &quot;, replacement = &quot;_&quot;) %&gt;% 
    tolower() %&gt;% 
    gsub(x = ., pattern = &quot;organism_&quot;, replacement = &quot;&quot;)


gold_reduced &lt;- gold %&gt;% 
    select(ncbi_tax_id, ncbi_superkingdom,  
            ncbi_phylum, ncbi_class, ncbi_order, ncbi_family, ncbi_genus, ncbi_species, 
            name, gram_stain, metabolism, oxygen_requirement, 
            sporulation, motility, cell_shape) %&gt;% 
    rename(&quot;species_tax_id&quot; = ncbi_tax_id,
           &quot;superkingdom&quot; = ncbi_superkingdom,
           &quot;phylum&quot; = ncbi_phylum,
           &quot;class&quot; = ncbi_class,
           &quot;order&quot; = ncbi_order,
           &quot;family&quot; = ncbi_family,
           &quot;genus&quot; = ncbi_genus,
           &quot;species&quot; = ncbi_species,
           &quot;pathways&quot; = metabolism,
           &quot;metabolism&quot; = oxygen_requirement) %&gt;% as.data.table()

# nest traits 
tbl &lt;- gold_reduced %&gt;%
    select(-name) %&gt;%
    group_by(species_tax_id, superkingdom, phylum, order, 
             family, genus, species) %&gt;%
    nest(traits = c(gram_stain, pathways, metabolism, 
           cell_shape, motility, sporulation))
    

# a subset of the table that has more than one row per trait nested values 
tbl_munge &lt;- tbl %&gt;% filter(map_lgl(traits, ~{nrow(.x) &gt; 1}))</code></pre>
<p>The issue is that there some species with multiple entries per
species due to differences in certain characteristics but also multiple
strains per species. Here, we’re going to loop over all of those rows
and then replace them. Since the original database is at species level,
we’re also going to collapse from the strain level to the species level
similar to Madin et al. 2020</p>
<p>Let’s define a function that processes species that have multiple
entries in GOLD</p>
<pre class="r"><code># This function takes a data frame and a column 
# and selects the response with the highest frequency
select_best &lt;- function(df, column){
    vec &lt;- unlist(df[,..column])
    freq &lt;- as.data.frame(table(vec))
    if (nrow(freq) == 0){
        return(NA_character_)
    } else {
        freq &lt;- freq %&gt;% mutate(prop = Freq/sum(Freq)) %&gt;%
            filter(prop &gt;= 0.5) %&gt;% top_n(n = 1, wt = prop)
        return(freq %&gt;% pull(vec) %&gt;% as.vector())
    }
}

# This function then utilizes select_best
# to process entries with duplicates (more than one row)
# for pathways, the goal is to concatenate them
process_duplicates &lt;- function(df){
    # get only unique rows
    df &lt;- unique(df)
    if (nrow(df) == 1){
        return(df)
    }
    v &lt;- c(&quot;gram_stain&quot;, &quot;pathways&quot;, &quot;metabolism&quot;, 
           &quot;sporulation&quot;, &quot;motility&quot;, &quot;cell_shape&quot;)
    suppressMessages(res &lt;- map_dfc(v, ~{
        if (.x == &quot;pathways&quot;){
            str_vec &lt;- na.omit(df$pathways) %&gt;% as.vector()
            if (length(str_vec) == 0){
                out &lt;- NA_character_
            } else {
                out &lt;- str_replace(str_vec, pattern = &quot; &quot;, 
                                   replacement = &quot;_&quot;) %&gt;% 
                    paste(collapse = &quot;, &quot;)
            }
        } else {
            out &lt;- select_best(df, .x)
        }
        return(out)
    }))
    colnames(res) &lt;- v
    res &lt;- as.data.table(res)
    return(res)
}</code></pre>
<p>Let’s apply it to <code>tbl_munge</code> and then merge both data
frames together into one consensus trait. After this munging, the final
GOLD database is processed and ready to merge into the master
<code>base</code> data frame</p>
<pre class="r"><code>
tbl_munge &lt;- tbl_munge %&gt;% 
    mutate(traits = map(traits, process_duplicates)) %&gt;% as_tibble() %&gt;%
    unnest(traits)</code></pre>
<p>We merge by extracting the <code>species_tax_id</code> column out of
<code>tbl</code>, remove all rows with that identifier and then replace
that with those from the <code>tbl_munge</code> database.</p>
<pre class="r"><code>
ids &lt;- tbl_munge %&gt;% pull(species_tax_id)

gold_final &lt;- tbl %&gt;% filter(!species_tax_id %in% ids) %&gt;% 
    as_tibble() %&gt;% 
    unnest(traits) %&gt;% bind_rows(tbl_munge)

# final cleaning
gold_final &lt;- gold_final %&gt;% 
    mutate(metabolism = tolower(metabolism), 
           gram_stain = if_else(gram_stain == &quot;Gram-&quot;, &quot;negative&quot;, &quot;positive&quot;), 
           sporulation = if_else(sporulation == &quot;Nonsporulating&quot;, &quot;no&quot;, &quot;yes&quot;), 
           motility = case_when(
               motility == &quot;Nonmotile&quot; ~ &quot;no&quot;, 
               motility == &quot;Motile&quot; ~ &quot;yes&quot;, 
               TRUE ~ motility
           ), 
           cell_shape = tolower(str_replace(cell_shape,&quot;-shaped&quot;,&quot;&quot;)),
           cell_shape = case_when(
               cell_shape %in% c(&quot;rod&quot;) ~ &quot;bacillus&quot;,
               cell_shape %in% c(&quot;sphere&quot;, &quot;oval&quot;, 
                                 &quot;bean&quot;, &quot;coccoid&quot;, &quot;ovoid&quot;, 
                                 &quot;spore&quot;) ~ &quot;coccus&quot;, 
               cell_shape %in% c(&quot;helical&quot;) ~ &quot;spiral&quot;, 
               cell_shape %in% c(&quot;curved&quot;) ~ &quot;vibrio&quot;, 
               cell_shape %in% c(&quot;flask&quot;, &quot;open-ring&quot;, &quot;lancet&quot;) ~ &quot;irregular&quot;, 
               # only Mycoplasma genitalium for flask and 
               # only Thiomicrospira cyclica for open-ring
               # only Nitrolancea hollandica for lancet
               TRUE ~ cell_shape
           ))

head(gold_final)
#&gt; # A tibble: 6 × 14
#&gt;   species_tax_id superkingdom phylum class order family genus species gram_stain
#&gt;            &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;     
#&gt; 1         515635 Bacteria     Dicty… Dict… Dict… Dicty… Dict… Dictyo… positive  
#&gt; 2         521011 Archaea      Eurya… Meth… Meth… Metha… Meth… Methan… positive  
#&gt; 3         498848 Bacteria     Deino… Dein… Ther… Therm… Ther… Thermu… negative  
#&gt; 4         481743 Bacteria     Firmi… Baci… Baci… Paeni… Paen… Paenib… positive  
#&gt; 5         634499 Bacteria     Prote… Gamm… Ente… Erwin… Erwi… Erwini… negative  
#&gt; 6         580327 Bacteria     Firmi… Clos… Ther… Therm… Ther… Thermo… positive  
#&gt; # … with 5 more variables: pathways &lt;chr&gt;, metabolism &lt;chr&gt;, sporulation &lt;chr&gt;,
#&gt; #   motility &lt;chr&gt;, cell_shape &lt;chr&gt;
# pray not to crash
rm(gold)
rm(gold_reduced)
rm(tbl)
rm(tbl_munge)
gc()
#&gt;           used (Mb) gc trigger  (Mb) limit (Mb) max used  (Mb)
#&gt; Ncells 1725958 92.2    5509228 294.3         NA  6886535 367.8
#&gt; Vcells 6699406 51.2   76208405 581.5      16384 95260506 726.8</code></pre>
<p><code>gold_final</code> will now be the finalized form of the GOLD
database as put into the original format of Madin et al. </p>
</div>
<div id="processing-weissman-et-al." class="section level3">
<h3>Processing Weissman et al. </h3>
<pre class="r"><code>
weissman &lt;- read_csv(here(&quot;data&quot;, &quot;weissman.csv&quot;))
#&gt; Rows: 3369 Columns: 174
#&gt; ── Column specification ────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr  (16): Organism, kingdom, phylum, class, order, family, genus, species, ...
#&gt; dbl (158): taxid_kingdom, taxid_phylum, taxid_class, taxid_order, taxid_fami...
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

# select the relevant columns
weissman &lt;- weissman %&gt;% select(c(&quot;taxid_species&quot;,
                                  &quot;kingdom&quot;, &quot;phylum&quot;, &quot;class&quot;, &quot;order&quot;, &quot;family&quot;, &quot;genus&quot;, &quot;species&quot;,
                                  &quot;Motility_general&quot;, &quot;Oxygen.Preference&quot;, &quot;Cell.Shape&quot;, &quot;Cell.Aggregation&quot;,
                                  starts_with(&quot;Enzyme.Assays&quot;), 
                                  starts_with(&quot;Volatile.Gas.Production&quot;), 
                                  starts_with(&quot;Substrate.Utilization&quot;)))</code></pre>
<p>Here, we’re going to re-format it similar to Madin et
al. <code>Enzyme.Assays</code> and <code>Volatile.Gas.Production</code>
is equivalent to <code>pathways</code> while
<code>Substrate.Utililization</code> is equivalent to
<code>carbon_substrates</code>. Since there might be non-carbon
compounds here, we’re going to rename Madin et al.’s
<code>carbon_substrates</code> into just <code>substrate</code> here
similar to Weissman’s database.</p>
<pre class="r"><code>weissman &lt;- weissman %&gt;% group_by(taxid_species, kingdom, phylum, 
                      class, order, family, genus, species) %&gt;%
    nest(pathways = starts_with(c(&quot;Enzyme.Assays&quot;, &quot;Volatile.Gas.Production&quot;)), 
         substrate = starts_with(&quot;Substrate.Utilization&quot;)) %&gt;% ungroup() %&gt;% 
    mutate(across(where(is.character), ~na_if(., &quot;0&quot;))) %&gt;% 
    rename(&quot;species_tax_id&quot; = taxid_species, &quot;superkingdom&quot; = kingdom, 
           &quot;metabolism&quot; = Oxygen.Preference,
           &quot;motility&quot; = Motility_general, &quot;cell_shape&quot; = Cell.Shape, &quot;cell_aggregation&quot; = Cell.Aggregation)

# motility, cell shape, metabolism
weissman$cell_shape %&gt;% unique()
#&gt; [1] NA                    &quot;rod&quot;                 &quot;ovoid/coccobacillus&quot;
#&gt; [4] &quot;spirillum/corkscrew&quot;

weissman &lt;- weissman %&gt;% 
    mutate(motility = if_else(motility == &quot;non-motile&quot;, &quot;yes&quot;, &quot;no&quot;), 
                    cell_shape = case_when(
                        cell_shape == &quot;rod&quot; ~ &quot;bacillus&quot;,
                        cell_shape == &quot;ovoid/coccobacillus&quot; ~ &quot;coccus&quot;, 
                        cell_shape == &quot;spirillum/corkscrew&quot; ~ &quot;spiral&quot;,
                        TRUE ~ cell_shape
                    ), 
                    metabolism = case_when(
                        metabolism == &quot;microaerophile&quot; ~ &quot;microaerophilic&quot;,
                        TRUE ~ metabolism
                    ), 
                    metabolism = str_replace(metabolism, 
                                             pattern = &quot;obe$&quot;, replacement = &quot;obic&quot;))</code></pre>
<p>Here we’re going to define some functions to process the traits that
is internal (collapse into one column) for pathways and for
substrates</p>
<pre class="r"><code>#&#39; @param unit A single unit from a list of trait data frames
proc_pathways &lt;- function(unit){
    string &lt;- unit %&gt;% pivot_longer(everything()) %&gt;% 
        filter(value != 0)
    if (nrow(string) == 0){
        return(NA)
    } else {
        string &lt;- string %&gt;% 
            mutate(name = str_replace(name, pattern = &quot;Enzyme.Assays..&quot;, replacement = &quot;&quot;)) %&gt;%
            mutate(name = str_replace(name, pattern = &quot;Volatile.Gas.Production..&quot;, replacement = &quot;synthesis_&quot;)) %&gt;%
            mutate(name = str_replace(name, pattern = &quot;\\.$&quot;, &quot;&quot;)) %&gt;%
            mutate(name = str_replace(name, pattern = &quot;\\.\\.(.*)$&quot;, &quot;&quot;)) %&gt;%
            mutate(name = str_replace_all(name, pattern = &quot;\\.&quot;, &quot;_&quot;)) %&gt;% 
            rowwise() %&gt;%
            mutate(name = if_else(str_detect(name, &quot;synthesis_&quot;), 
                                  true = paste(rev(str_split(name, pattern = &quot;_&quot;, 
                                                             n = 2)[[1]]), 
                                               collapse = &quot;_&quot;),
                                  false = name)) %&gt;% 
            ungroup() %&gt;%
            pull(name) %&gt;% paste(., collapse = &quot;, &quot;)
    }
        
    return(string)
}

proc_substrate &lt;- function(unit){
    string &lt;- unit %&gt;% pivot_longer(everything()) %&gt;% filter(value != 0)
    if (nrow(string) == 0){
        return(NA)
    } else {
        string &lt;- string %&gt;% mutate(name = str_split(name, pattern = &quot;\\.\\.&quot;, 
                                           n = 2, simplify = TRUE)[,2]) %&gt;%
            mutate(name = str_replace(name, &quot;(\\.\\.|\\.)$&quot;, &quot;&quot;)) %&gt;% 
            mutate(name = str_replace_all(name, &quot;(\\.\\.|\\.)&quot;, &quot;_&quot;)) %&gt;% 
            pull(name) %&gt;% paste(., collapse = &quot;, &quot;)
    }
    return(string)
}</code></pre>
<p>Map it over the two columns in the original data frame.</p>
<pre class="r"><code>substr &lt;- map_chr(weissman$substrate, proc_substrate)
pthway &lt;- map_chr(weissman$pathways, proc_pathways)

weissman &lt;- weissman %&gt;% select(-c(pathways, substrate)) %&gt;% 
  mutate(substrate = substr, pathways = pthway)</code></pre>
</div>
<div id="checking-for-duplicates" class="section level3">
<h3>Checking for duplicates</h3>
<p>First, we define a function to check for similar sounding names
across all the unique pathways and substrates for all the data sets.
Here, we use the <code>stringdist</code> function from the
<code>stringdist</code> package. We use the standard OSA metric (also
called the Damerau-Levenshtein distance) to query for potential
similarly sounding names of identical pathways or compounds.</p>
<pre class="r"><code>
check_matches &lt;- function(df, type=c(&quot;pathways&quot;,&quot;substrates&quot;)){
    b_val &lt;- base %&gt;% pull(!!type) %&gt;% unique() %&gt;% str_split(pattern = &quot;, &quot;) %&gt;%
        unlist() %&gt;% unique() %&gt;% na.omit() %&gt;% as.vector()
    
    q_val &lt;- df %&gt;% pull(!!type) %&gt;% unique() %&gt;% str_split(pattern = &quot;, &quot;) %&gt;% 
        unlist() %&gt;% unique() %&gt;% na.omit() %&gt;% as.vector()
    
    check &lt;- map(q_val, ~{
        match &lt;- stringdist(a = .x, b = b_val)
        # match 0 is the same, and match &gt; 2 is too different 
        ret &lt;- b_val[match &gt; 0 &amp; match &lt;= 2]
        if (length(ret) == 0){
            return(NA)
        } else {
            out &lt;- tibble(
                query = rep(.x, length(ret)),
                ref = ret
            )
        }
    })
    check &lt;- check[!sapply(check, function(x) all(is.na(x)))]
    
    return(check)
}</code></pre>
<p>Check Weissman et al. database</p>
<pre class="r"><code>Reduce(check_matches(weissman, &quot;pathways&quot;), f = rbind)
#&gt; # A tibble: 2 × 2
#&gt;   query               ref                
#&gt;   &lt;chr&gt;               &lt;chr&gt;              
#&gt; 1 elastin_degradation plastic_degradation
#&gt; 2 elastin_degradation gelatin_degradation
Reduce(check_matches(weissman, &quot;substrate&quot;), f = rbind)
#&gt; # A tibble: 55 × 2
#&gt;    query    ref      
#&gt;    &lt;chr&gt;    &lt;chr&gt;    
#&gt;  1 sucrose  fucose   
#&gt;  2 butanol  2-butanol
#&gt;  3 butanol  1-butanol
#&gt;  4 ethanol  methanol 
#&gt;  5 glucose  fucose   
#&gt;  6 mannose  maltose  
#&gt;  7 fructose fucose   
#&gt;  8 lactose  lactate  
#&gt;  9 lactose  galactose
#&gt; 10 lactose  maltose  
#&gt; # … with 45 more rows</code></pre>
<p>We can see that for a lot of the compounds the names might be the
same but they’re actually different. However, there are certain
conventions such as “_” for spaces or “-” that we might need to address
for the final merge.</p>
<p>Let’s check the GOLD database for similar sounding names. Since GOLD
does not have substrate information, we only check for pathways</p>
<pre class="r"><code>Reduce(check_matches(gold_final, &quot;pathways&quot;), f = rbind)
#&gt; # A tibble: 6 × 2
#&gt;   query              ref               
#&gt;   &lt;chr&gt;              &lt;chr&gt;             
#&gt; 1 Nitrogen fixation  nitrogen_fixation 
#&gt; 2 Methane oxidation  methane_oxidation 
#&gt; 3 Chitin degradation chitin_degradation
#&gt; 4 Chitin_degradation chitin_degradation
#&gt; 5 Nitrogen_fixation  nitrogen_fixation 
#&gt; 6 Methane_oxidation  methane_oxidation</code></pre>
<p>We can see a similar pattern here. Essentially we need to wrangle the
columns such that spaces are replaced with underscores and all names
should be to lower case.</p>
<pre class="r"><code>trim_path &lt;- function(vec){
    vec &lt;- vec %&gt;% tolower() %&gt;% 
        str_split(pattern = &quot;(, |\\|)&quot;) %&gt;% 
        map(~{
            str_trim(.x) %&gt;% str_replace_all(&quot;\\-&quot;, &quot;&quot;) %&gt;%
                str_replace_all(&quot; &quot;, &quot;_&quot;) %&gt;% unique()
        })
    return(vec)
}


gold_final$pathways &lt;- trim_path(gold_final$pathways)
base$pathways &lt;- trim_path(base$pathways)
weissman$pathways &lt;- trim_path(weissman$pathways)
weissman$substrate &lt;- trim_path(weissman$substrate)
base$substrate &lt;- trim_path(base$substrate)</code></pre>
</div>
<div id="combine-all-data-frames" class="section level3">
<h3>Combine all data frames</h3>
<p>After munging, let’s combine all of the names! The strategy is very
similar to handling multiple entries for GOLD. First, we bind all of our
databases together. Then, we <code>group_by</code> and nest all our
trait data into a list. Then we process these lists and return
<code>unique</code> rows (deduplicated). If the rows are not unique,
then we process the non-unique rows by either concatenating the traits
together or vote on consensus using the top most represented trait.</p>
<pre class="r"><code>complete &lt;- bind_rows(
    base %&gt;% mutate(source = &quot;madin&quot;), 
    gold_final %&gt;% mutate(source = &quot;gold&quot;), 
    weissman %&gt;% mutate(source = &quot;weissman&quot;)
)

tally &lt;- complete %&gt;% filter(!is.na(species)) %&gt;% 
    group_by(species_tax_id, superkingdom, phylum, class, order, 
             family, genus, species) %&gt;% 
    tally()


multiple_rows &lt;- tally %&gt;% filter(n &gt;= 2) %&gt;% pull(species_tax_id)

reconcile &lt;- complete %&gt;% filter(species_tax_id %in% multiple_rows)

reconcile &lt;- reconcile %&gt;% group_by(species_tax_id, superkingdom, phylum, class, order,
                       family, genus, species, metabolism) %&gt;%
    nest(data = c(metabolism, gram_stain, pathways, substrate, sporulation, motility, cell_shape, cell_aggregation, source)) %&gt;% ungroup()</code></pre>
<p>Similar to above, let’s write a function that goes through each
element and reconcile the different sources.</p>
<pre class="r"><code>#&#39; @param df This is a data frame of multiple columns, where the columns of 
#&#39;     pathways and substrates are themselves lists 
collapse_trait &lt;- function(df){
    nonlist &lt;- c(&quot;metabolism&quot;, &quot;gram_stain&quot;, &quot;sporulation&quot;, &quot;motility&quot;, 
                 &quot;cell_shape&quot;, &quot;cell_aggregation&quot;)    
    out &lt;- suppressMessages(map_dfc(nonlist, ~{
        traits &lt;- df %&gt;% pull(.x)
        traits &lt;- as.data.frame(table(traits))
        if (nrow(traits) &gt;= 1){
            return(traits %&gt;% 
                       mutate(prop = Freq/sum(Freq)) %&gt;%
                       filter(prop == max(prop) &amp; prop &gt;= 0.5) %&gt;% 
                       pull(traits) %&gt;% as.vector()
            )
        } else {
            return(NA_character_)
        }
    }))
    names(out) &lt;- nonlist
    out$pathways &lt;- list(df %&gt;% pull(&quot;pathways&quot;) %&gt;% 
                             Reduce(f = c, x = .) %&gt;% unique())
    out$substrate &lt;- list(df %&gt;% pull(&quot;substrate&quot;) %&gt;% 
                              Reduce(f = c, x = .) %&gt;% 
                              unique())
    return(out)
}

reconcile &lt;- reconcile %&gt;% mutate(traits = map(data, collapse_trait)) %&gt;% 
    select(-data)</code></pre>
<p>Let’s merge the collapsed file and save everything to a database</p>
<pre class="r"><code>complete &lt;- bind_rows(reconcile %&gt;% unnest(traits), 
          complete %&gt;% filter(!species_tax_id %in% multiple_rows))

saveRDS(complete, file = here(&quot;output&quot;, &quot;db_merged.rds&quot;))
complete %&gt;% mutate(
    pathways = map_chr(pathways, ~{
        paste(.x, collapse = &quot;, &quot;)
    }), 
    substrate = map_chr(substrate, ~{
        paste(.x, collapse = &quot;, &quot;)
    })
) %&gt;% write_csv(x = ., file = here(&quot;output&quot;, &quot;db_merged.csv&quot;))</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
#&gt; R version 4.1.2 (2021-11-01)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Big Sur 10.16
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] forcats_0.5.1     stringr_1.4.0     dplyr_1.0.8       purrr_0.3.4      
#&gt;  [5] readr_2.1.2       tidyr_1.2.0       tibble_3.1.6      ggplot2_3.3.5    
#&gt;  [9] tidyverse_1.3.1   stringdist_0.9.8  here_1.0.1        targets_0.10.0   
#&gt; [13] dtplyr_1.2.1      data.table_1.14.2 piggyback_0.1.1  
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] fs_1.5.2            bit64_4.0.5         lubridate_1.8.0    
#&gt;  [4] httr_1.4.2          rprojroot_2.0.2     gh_1.3.0           
#&gt;  [7] tools_4.1.2         backports_1.4.1     bslib_0.3.1        
#&gt; [10] utf8_1.2.2          R6_2.5.1            DBI_1.1.2          
#&gt; [13] colorspace_2.0-3    withr_2.5.0         tidyselect_1.1.2   
#&gt; [16] processx_3.5.2      bit_4.0.4           compiler_4.1.2     
#&gt; [19] git2r_0.29.0        cli_3.2.0           rvest_1.0.2        
#&gt; [22] xml2_1.3.3          sass_0.4.0          scales_1.1.1       
#&gt; [25] callr_3.7.0         digest_0.6.29       rmarkdown_2.12     
#&gt; [28] pkgconfig_2.0.3     htmltools_0.5.2     dbplyr_2.1.1       
#&gt; [31] fastmap_1.1.0       rlang_1.0.2         readxl_1.3.1       
#&gt; [34] rstudioapi_0.13     jquerylib_0.1.4     generics_0.1.2     
#&gt; [37] jsonlite_1.8.0      vroom_1.5.7         magrittr_2.0.2     
#&gt; [40] Rcpp_1.0.8          munsell_0.5.0       fansi_1.0.2        
#&gt; [43] lifecycle_1.0.1     stringi_1.7.6       whisker_0.4        
#&gt; [46] yaml_2.3.5          grid_4.1.2          parallel_4.1.2     
#&gt; [49] promises_1.2.0.1    crayon_1.5.0        haven_2.4.3        
#&gt; [52] hms_1.1.1           knitr_1.37          ps_1.6.0           
#&gt; [55] pillar_1.7.0        igraph_1.2.11       base64url_1.4      
#&gt; [58] codetools_0.2-18    clisymbols_1.2.0    reprex_2.0.1       
#&gt; [61] glue_1.6.2          evaluate_0.15       renv_0.15.4        
#&gt; [64] BiocManager_1.30.16 modelr_0.1.8        vctrs_0.3.8        
#&gt; [67] tzdb_0.2.0          httpuv_1.6.5        cellranger_1.1.0   
#&gt; [70] gtable_0.3.0        assertthat_0.2.1    cachem_1.0.6       
#&gt; [73] xfun_0.30           broom_0.7.12        later_1.3.0        
#&gt; [76] memoise_2.0.1       workflowr_1.7.0     ellipsis_0.3.2</code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
