<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Quang Nguyen" />

<meta name="date" content="2022-03-12" />

<title>Preparing trait databases</title>

<script src="site_libs/header-attrs-2.12/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Evaluating trait databases for taxon-set analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Supplemental
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01_retrieve_agp.html">Downloading AGP data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/qpmnguyen/microbe_set_trait">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preparing trait databases</h1>
<h4 class="author">Quang Nguyen</h4>
<h4 class="date">2022-03-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-03-14
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>microbe_set_trait/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220309code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220309)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220309code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220309)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomqpmnguyenmicrobesettraittree7175fe96505114d870bc34e069628e1aa6e259d8targetblank7175fe9a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/qpmnguyen/microbe_set_trait/tree/7175fe96505114d870bc34e069628e1aa6e259d8" target="_blank">7175fe9</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomqpmnguyenmicrobesettraittree7175fe96505114d870bc34e069628e1aa6e259d8targetblank7175fe9a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/qpmnguyen/microbe_set_trait/tree/7175fe96505114d870bc34e069628e1aa6e259d8" target="_blank">7175fe9</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Renviron
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .nextflow.log
    Ignored:    .nextflow.log.1
    Ignored:    .nextflow.log.2
    Ignored:    .nextflow.log.3
    Ignored:    .nextflow.log.4
    Ignored:    .nextflow.log.5
    Ignored:    .nextflow.log.6
    Ignored:    .nextflow/
    Ignored:    R/prototype.R
    Ignored:    _targets/
    Ignored:    databases/
    Ignored:    large_files/
    Ignored:    picrust2_tmp/
    Ignored:    python/.ipynb_checkpoints/
    Ignored:    python/data/
    Ignored:    renv/staging/
    Ignored:    store_coverage/
    Ignored:    store_dada2/
    Ignored:    store_db_preprocess/
    Ignored:    store_pred/
    Ignored:    work/

Unstaged changes:
    Modified:   analysis/db_prep.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/db_prep.Rmd</code>) and HTML (<code>docs/db_prep.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/7175fe96505114d870bc34e069628e1aa6e259d8/analysis/db_prep.Rmd" target="_blank">7175fe9</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-14
</td>
<td>
Add a skip function
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/qpmnguyen/microbe_set_trait/d785d21db27e033600d31dbc5ea2b4dfac63f60f/docs/db_prep.html" target="_blank">d785d21</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/qpmnguyen/microbe_set_trait/e73f8c0d32289aafbcb30b15c6862f61c042b26a/docs/db_prep.html" target="_blank">e73f8c0</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/qpmnguyen/microbe_set_trait/blob/909f189e3b47953c57df3cbcb456cec0d0678917/analysis/db_prep.Rmd" target="_blank">909f189</a>
</td>
<td>
Quang Nguyen
</td>
<td>
2022-03-12
</td>
<td>
workflowr::wflow_publish(“analysis/db_prep.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>We’re prepping the latest database for export for evaluation. For this manuscript, we’re merging a couple of existing databases:</p>
<ol style="list-style-type: decimal">
<li>The comprehensive synthesis of trait-database from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7275036/">Madin et al. 2020</a>. This database was last updated in 2020. Most of the database’s sources are static sources, with the exception of the <a href="https://gold.jgi.doe.gov/downloads">GOLD database</a>. As such, we’re merging the existing release of the Madin et al. database with the most recent GOLD release (2022-03-12).<br />
</li>
<li>Manual curation of bergey’s manual by <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04216-2">Weissman et al.</a>. This database contains manual curation of the Bergey’s manual specific to human-associated microbiomes.</li>
</ol>
<p>The way we’re trying to combine these disparate sources would be to perform something similar to Madin et al. using the <a href="https://github.com/bacteria-archaea-traits/bacteria-archaea-traits/blob/master/R/functions.R">R code</a> on GitHub. The step-by-step workflow is:</p>
<ol style="list-style-type: decimal">
<li>Prepare and process each individual database file.<br />
</li>
<li>Second, concatenate all database files by row. Remember to specify data source.</li>
<li></li>
</ol>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<p>Loading some packages</p>
<pre class="r"><code>library(piggyback)
library(data.table)
library(dtplyr)
library(targets)
library(here)
#&gt; here() starts at /Users/quangnguyen/research/microbe_set_trait
library(tidyverse)
#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──
#&gt; ✓ ggplot2 3.3.5     ✓ purrr   0.3.4
#&gt; ✓ tibble  3.1.6     ✓ dplyr   1.0.8
#&gt; ✓ tidyr   1.2.0     ✓ stringr 1.4.0
#&gt; ✓ readr   2.1.2     ✓ forcats 0.5.1
#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
#&gt; x dplyr::between()   masks data.table::between()
#&gt; x dplyr::filter()    masks stats::filter()
#&gt; x dplyr::first()     masks data.table::first()
#&gt; x dplyr::lag()       masks stats::lag()
#&gt; x dplyr::last()      masks data.table::last()
#&gt; x purrr::transpose() masks data.table::transpose()
here::i_am(&quot;analysis/db_prep.Rmd&quot;)
#&gt; here() starts at /Users/quangnguyen/research/microbe_set_trait</code></pre>
<div id="downloading-data-files" class="section level3">
<h3>Downloading data files</h3>
<p>Downloading data files from respective sources and version control on GitHub using <code>piggyback</code></p>
<pre class="r"><code># Not run since file is large and is already uploaded to GitHub. Code here for provenance 
piggyback::pb_upload(file = here(&quot;large_files&quot;, &quot;goldData.xlsx&quot;), tag = &quot;0.1&quot;, overwrite = TRUE)
</code></pre>
</div>
<div id="processing-data-files" class="section level3">
<h3>Processing data files</h3>
<p>First, load the original Madin et al. database</p>
<pre class="r"><code>base &lt;- read_csv(here(&quot;data&quot;, &quot;condensed_species_NCBI.txt&quot;))
#&gt; Rows: 14893 Columns: 79
#&gt; ── Column specification ────────────────────────────────────────────────────────
#&gt; Delimiter: &quot;,&quot;
#&gt; chr (19): species, genus, family, order, class, phylum, superkingdom, gram_s...
#&gt; dbl (60): species_tax_id, d1_lo, d1_up, d2_lo, d2_up, doubling_h, genome_siz...
#&gt; 
#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.
#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div id="processing-gold" class="section level4">
<h4>Processing GOLD</h4>
<p>Let’s process GOLD. First, load data from <code>piggyback</code> (this is cached). Script follows existing processing pipeline <a href="https://github.com/bacteria-archaea-traits/bacteria-archaea-traits/blob/master/R/preparation/gold.R">here</a> from Madin et al. </p>
<pre class="r"><code>pth &lt;- here(&quot;large_files&quot;)
if (!file.exists(pth)){
    piggyback::pb_download(file = &quot;goldData.xlsx&quot;, dest = pth, tag = &quot;0.1&quot;)
}
gold &lt;- readxl::read_xlsx(path = file.path(pth, &quot;goldData.xlsx&quot;), sheet = &quot;Organism&quot;)
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM4230 / R4230C39: got &#39;Phenanthrene, Formic acid, gluconic
#&gt; acid, malonic acid&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM6852 / R6852C39: got &#39;CO2, Acetate&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM10492 / R10492C39: got &#39;Sweet sugar&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM11601 / R11601C39: got &#39;d-fructose, d-galactose, d-
#&gt; glucose, lactose, raffinose and sucrose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM11931 / R11931C39: got &#39;Carbohydrates, peptides&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM11950 / R11950C39: got &#39;Mono- and polysaccharides&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM12150 / R12150C39: got &#39;Glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM16410 / R16410C39: got &#39;Heterotroph (galactose, glucose,
#&gt; tryptic soy broth, fructose, yeast extract)&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM27499 / R27499C39: got &#39;Glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM28889 / R28889C39: got &#39;Sucrose, glucose, fructose,
#&gt; galactose, maltose, mannitol, mannose, arabitol&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM33787 / R33787C39: got &#39;carbohydrates, amino-acids,
#&gt; organic acid anions and alcohols&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM35459 / R35459C39: got &#39;polycyclic aromatic hydrocarbons&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM35962 / R35962C39: got &#39;Heterotroph&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM52183 / R52183C39: got &#39;Fructose, lactose, acetate&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM59612 / R59612C39: got &#39;Sucrose, glucose, fructose,
#&gt; galactose, maltose, mannitol, mannose, arabitol&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM63996 / R63996C39: got &#39;Glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM66310 / R66310C39: got &#39;amino acids&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK83277 / R83277C37: got &#39;0–30 MPa (Optimal 10MPa)&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK84355 / R84355C37: got &#39;normal&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK228232 / R228232C37: got &#39;10 MPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK240466 / R240466C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK266739 / R266739C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK267650 / R267650C37: got &#39;107.55.82&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK267717 / R267717C37: got &#39;107.55.82&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK267719 / R267719C37: got &#39;107.55.82&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274564 / R274564C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274565 / R274565C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274566 / R274566C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274567 / R274567C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274568 / R274568C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274569 / R274569C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274570 / R274570C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274571 / R274571C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274572 / R274572C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274573 / R274573C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274574 / R274574C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274575 / R274575C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274576 / R274576C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274577 / R274577C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274578 / R274578C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274579 / R274579C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274580 / R274580C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274581 / R274581C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274621 / R274621C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274622 / R274622C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274623 / R274623C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274624 / R274624C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274625 / R274625C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274626 / R274626C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274627 / R274627C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274628 / R274628C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274629 / R274629C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274630 / R274630C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274631 / R274631C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274632 / R274632C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274633 / R274633C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274634 / R274634C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274635 / R274635C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274636 / R274636C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274637 / R274637C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274638 / R274638C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274639 / R274639C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274640 / R274640C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274641 / R274641C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274642 / R274642C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274643 / R274643C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274644 / R274644C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274645 / R274645C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274646 / R274646C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274647 / R274647C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274648 / R274648C37: got &#39;14532 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK274649 / R274649C37: got &#39;101.3 kPa&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK279601 / R279601C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK283545 / R283545C37: got &#39;atmospheric pressure&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM312483 / R312483C39: got &#39;CO, fructose, galactose,
#&gt; glucose, mannose, raffinose, ribose, sucrose &amp; xylose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM312514 / R312514C39: got &#39;citrate, cellobiose, malate,
#&gt; fumarate, malonate&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM332160 / R332160C39: got &#39;cellobiose, l-fucose, dextrin,
#&gt; lactose, d-galactose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM333300 / R333300C39: got &#39;d-glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM338846 / R338846C39: got &#39;Urea, Raffinose, p-Hydroxy-
#&gt; phenylacetic acid, Methyl-pyruvate, l-Galactonic acid lactone&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM343777 / R343777C39: got &#39;methanol, formate&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK360146 / R360146C37: got &#39;1atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK360147 / R360147C37: got &#39;1atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK362424 / R362424C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM365586 / R365586C39: got &#39;Dextrin, Gentiobiose, Pectin&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM365658 / R365658C39: got &#39;d-Mannitol, d-Sorbitol&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM372436 / R372436C39: got &#39;glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM373837 / R373837C39: got &#39;cellobiose, fructose, maltose,
#&gt; dextrin, N-acetylglucosamine, gluconate and glycerol&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK375118 / R375118C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK375119 / R375119C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK375120 / R375120C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK375121 / R375121C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AK375122 / R375122C37: got &#39;1 atm&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM375867 / R375867C39: got &#39;Trehalose, Cellobiose,
#&gt; Turanose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM375889 / R375889C39: got &#39;Cellobiose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM376213 / R376213C39: got &#39;dextrin, glucose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM376238 / R376238C39: got &#39;Raffinose, Lactose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM378962 / R378962C39: got &#39;glucose, lactose, maltose and
#&gt; mannose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM380421 / R380421C39: got &#39;d-Galactose, d-glucose,
#&gt; mannitol, ribose, sucrose, trehalose and xylan&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM382545 / R382545C39: got &#39;d-Glucose, Mannose&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i =
#&gt; sheet, : Expecting logical in AM383476 / R383476C39: got &#39;d-fructose, maltose,
#&gt; cellobiose, sucrose, raffinose and l-asparagine&#39;
#&gt; Warning in read_fun(path = enc2native(normalizePath(path)), sheet_i = sheet, :
#&gt; Expecting logical in AM383720 / R383720C39: got &#39;d-Fructose, lactose and
#&gt; maltose&#39;


# convert names 
colnames(gold) &lt;- colnames(gold) %&gt;% 
    gsub(x = ., pattern = &quot; &quot;, replacement = &quot;_&quot;) %&gt;% 
    tolower() %&gt;% 
    gsub(x = ., pattern = &quot;organism_&quot;, replacement = &quot;&quot;)


gold_reduced &lt;- gold %&gt;% 
    select(ncbi_tax_id, ncbi_superkingdom,  
            ncbi_phylum, ncbi_order, ncbi_family, ncbi_genus, ncbi_species, 
            name, gram_stain, metabolism, oxygen_requirement, 
            sporulation, motility, cell_shape) %&gt;% 
    rename(&quot;species_tax_id&quot; = ncbi_tax_id,
           &quot;superkingdom&quot; = ncbi_superkingdom,
           &quot;phylum&quot; = ncbi_phylum,
           &quot;order&quot; = ncbi_order,
           &quot;family&quot; = ncbi_family,
           &quot;genus&quot; = ncbi_genus,
           &quot;species&quot; = ncbi_species,
           &quot;pathways&quot; = metabolism,
           &quot;metabolism&quot; = oxygen_requirement) %&gt;% as.data.table()
    

tbl &lt;- gold_reduced[1:1000,] %&gt;% as_tibble() %&gt;%
    group_by(species_tax_id, superkingdom, phylum, order, 
             family, genus, species, name) %&gt;%
    nest(traits = c(gram_stain, pathways, metabolism, 
           cell_shape, motility, sporulation))

tbl_munge &lt;- tbl %&gt;% filter(map_lgl(traits, ~{nrow(.x) &gt; 1}))

tbl_munge
#&gt; # A tibble: 2 × 9
#&gt; # Groups:   species_tax_id, superkingdom, phylum, order, family, genus,
#&gt; #   species, name [2]
#&gt;   species_tax_id superkingdom phylum   order family genus species name  traits  
#&gt;   &lt;chr&gt;          &lt;chr&gt;        &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt; &lt;list&gt;  
#&gt; 1 5011           Eukaryota    Ascomyc… Taph… Taphr… Taph… Taphri… Taph… &lt;tibble&gt;
#&gt; 2 62977          Bacteria     Proteob… Mora… Morax… Acin… Acinet… Acin… &lt;tibble&gt;</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but it won’t be re-run if previous chunks it depends on are updated. If you need to use caching, it is highly recommended to also set <code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the file (in a chunk that is not cached). Alternatively, you can customize the option <code>dependson</code> for each individual chunk that is cached. Using either <code>autodep</code> or <code>dependson</code> will remove this warning. See the <a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
<p>Let’s define a function that processes species that have multiple entries in GOLD</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
#&gt; R version 4.1.2 (2021-11-01)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Big Sur 10.16
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] forcats_0.5.1     stringr_1.4.0     dplyr_1.0.8       purrr_0.3.4      
#&gt;  [5] readr_2.1.2       tidyr_1.2.0       tibble_3.1.6      ggplot2_3.3.5    
#&gt;  [9] tidyverse_1.3.1   here_1.0.1        targets_0.10.0    dtplyr_1.2.1     
#&gt; [13] data.table_1.14.2 piggyback_0.1.1  
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] fs_1.5.2            bit64_4.0.5         lubridate_1.8.0    
#&gt;  [4] httr_1.4.2          rprojroot_2.0.2     gh_1.3.0           
#&gt;  [7] tools_4.1.2         backports_1.4.1     bslib_0.3.1        
#&gt; [10] utf8_1.2.2          R6_2.5.1            DBI_1.1.2          
#&gt; [13] colorspace_2.0-3    withr_2.5.0         tidyselect_1.1.2   
#&gt; [16] processx_3.5.2      bit_4.0.4           compiler_4.1.2     
#&gt; [19] git2r_0.29.0        cli_3.2.0           rvest_1.0.2        
#&gt; [22] xml2_1.3.3          sass_0.4.0          scales_1.1.1       
#&gt; [25] callr_3.7.0         digest_0.6.29       rmarkdown_2.12     
#&gt; [28] pkgconfig_2.0.3     htmltools_0.5.2     dbplyr_2.1.1       
#&gt; [31] fastmap_1.1.0       rlang_1.0.2         readxl_1.3.1       
#&gt; [34] rstudioapi_0.13     jquerylib_0.1.4     generics_0.1.2     
#&gt; [37] jsonlite_1.8.0      vroom_1.5.7         magrittr_2.0.2     
#&gt; [40] Rcpp_1.0.8          munsell_0.5.0       fansi_1.0.2        
#&gt; [43] lifecycle_1.0.1     stringi_1.7.6       whisker_0.4        
#&gt; [46] yaml_2.3.5          grid_4.1.2          parallel_4.1.2     
#&gt; [49] promises_1.2.0.1    crayon_1.5.0        haven_2.4.3        
#&gt; [52] hms_1.1.1           knitr_1.37          ps_1.6.0           
#&gt; [55] pillar_1.7.0        igraph_1.2.11       base64url_1.4      
#&gt; [58] codetools_0.2-18    clisymbols_1.2.0    reprex_2.0.1       
#&gt; [61] glue_1.6.2          evaluate_0.15       renv_0.15.4        
#&gt; [64] BiocManager_1.30.16 modelr_0.1.8        vctrs_0.3.8        
#&gt; [67] tzdb_0.2.0          httpuv_1.6.5        cellranger_1.1.0   
#&gt; [70] gtable_0.3.0        assertthat_0.2.1    cachem_1.0.6       
#&gt; [73] xfun_0.30           broom_0.7.12        later_1.3.0        
#&gt; [76] memoise_2.0.1       workflowr_1.7.0     ellipsis_0.3.2</code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
